### DISCUSSION TITLE
When binding a Object that implements INotifyCollectionChanged to a ItemsControl, update the object doesn't trigger view updating

### DISCUSSION THREAD

**Discussion by cdytoby** [AvaloniaUI/Avalonia#15623] [Q&A] [answered]
### Describe the bug

- In ItemsControl, ItemsSource is binded to a custom object that implments INotifyCollectionChanged, IEnumerable and IReadOnlyCollection
- that custom object gets updated, and CollectionChanged event is fired
- In debugger, the values of the custom object are all correct
- Bug: The actual ItemsControl view is empty

### To Reproduce

<s>
[DumbRenamer.zip](https://github.com/AvaloniaUI/Avalonia/files/15210435/DumbRenamer.zip)

This is my project (just started not long ago, to experiment things)

Just build and run the project: DumbRenamer.AvaloniaApp.Desktop

In file MainViewModel.cs, line 64, in the method FileInputViewOnCollectionChanged(), you can put a break point there, and this break point gets triggered.
</s>

Update:
I created a public repo for bug reporting (I think this probably won't be the last one).
https://github.com/cdytoby/AvaloniaBugReport/tree/bug/observable-list

For this bug specifically, it is the branch "bug/observable-list".

Just build and run the project "AvaloniaBugReport"

In file MainWindowViewModel.cs, line 37, method TestNotifiedEnumerableOnCollectionChanged(), you can set breakpoint there, it's the proof that INotifyCollectionChanged is called.

### Expected behavior

The ItemsControl displays the items correctly

### Avalonia version

11.0.10

### OS

Windows

### Additional context

I need a better ObservableCollection, so I found a third party one, they have their own solutions and I'd like to use it. 
https://github.com/Cysharp/ObservableCollections

To my understanding of MVVM, as long as INotifyCollectionChanged is correctly fired, view should be updated. (experience from WPF)

(I could use Avalonia ones but I don't want avalonia package involved when design view model. And the Avalonia.Collection namespace is in the Avalonia main package, it's not a independent package.)

**Comment by stevemonaco** [comment]
INCC needs to be on the UI thread. Using their `ObservableList<T>` works fine. It's only an issue with the `INotifyCollectionChangedSynchronizedView<T>`. Their WPF sample shows using `BindingOperations.EnableCollectionSynchronization`, so not being on the UI thread seems to be a possible scenario. I would have expected an exception or log message on Verbose, but I don't seem to get either.

**Comment by cdytoby** [comment]
I saw it, but I don't need to synchronize across thread (at least not right now), so I thought this is optional. `ObservableList<T>` don't have INCC, so `INotifyCollectionChangedSynchronizedView<T>` is the only option here if I want to keep using this.

**Comment by timunie** [comment]
Can you please file a minimum sample? I'll not explore a random file renamer as that could be harmful to my machine.

**Comment by cdytoby** [comment]
Hi, this name "DumbRenamer" is a project I just started, despite the solution name, there is not much code, and not much going on, that's why I upload it, and it's without any dll or built binary. 

But I understand your concern. Here are the some sample code files in a zip (.cs file and .axaml files are not directly allowed somehow)

[Code.zip](https://github.com/AvaloniaUI/Avalonia/files/15213876/Code.zip)

PS: CommunityToolkit.Mvvm package and ObservableCollections package from "https://github.com/Cysharp/ObservableCollections" are required.

**Reply by timunie** [reply]
If you want us to investigate an issue, we need a very minimal repo to understand that issue. Not a full app, not some fragments (fragments are okay if the problem is reproducible and easy to understand). I still believe it is an async issue as stated below.

**Comment by timunie** [comment]
You need the Dispather if you update things from a Task.

```cs
private async Task ChangeValue()
{
    await Task.Delay(1000);
    Dispatcher.UiThread.Post(() => { TestInt = 10 });
    ExecutePreview();
}
```

**Reply by cdytoby** [reply]
I just tried this, but it still doesn't work. 

In the project I mentioned, the new items in the list are still not appear in the UI.

**Comment by cdytoby** [comment]
Hi @timunie and everyone,

I created a public repo for bug reporting (I think this probably won't be the last one).

https://github.com/cdytoby/AvaloniaBugReport/tree/bug/observable-list

For this bug specifically, it is the branch "bug/observable-list".

I put the information to the first post too.

**Reply by timunie** [reply, marked as answer]
the docs state this for WPF: https://github.com/Cysharp/ObservableCollections?tab=readme-ov-file#wpf 

```
BindingOperations.EnableCollectionSynchronization(ItemsView, new object()); // for ui synchronization safety of viewmodel
```

So you should use another approach, for example DynamicData. It's by design with this lib.

**Reply by cdytoby** [reply]
OK, I see, thanks for the answer. I'll look into other approaches.
